src/components/DistanceRequest.js

```diff
@@ -197,62 +186,69 @@ function DistanceRequest({transactionID, report, transaction, mapboxAccessToken,
-                        let descriptionKey = 'distance.waypointDescription.';
-                        let waypointIcon;
-                        if (index === 0) {
-                            descriptionKey += 'start';
-                            waypointIcon = Expensicons.DotIndicatorUnfilled;
-                        } else if (index === lastWaypointIndex) {
-                            descriptionKey += 'finish';
-                            waypointIcon = Expensicons.Location;
-                        } else {
-                            descriptionKey += 'stop';
-                            waypointIcon = Expensicons.DotIndicator;
-                        }
-
-                        return (
-                            <MenuItemWithTopDescription
-                                description={translate(descriptionKey)}
-                                title={lodashGet(waypoints, [`waypoint${index}`, 'address'], '')}
-                                iconFill={theme.icon}
-                                secondaryIcon={waypointIcon}
-                                secondaryIconFill={theme.icon}
-                                shouldShowRightIcon
-                                onPress={() => navigateToWaypointEditPage(index)}
-                                key={key}
-                            />
-                        );
-                    })}
-                </ScrollView>
-                {shouldShowGradient && (
-                    <LinearGradient
-                        style={[styles.pAbsolute, styles.b0, styles.l0, styles.r0, {height: halfMenuItemHeight}]}
-                        colors={[StyleUtils.getTransparentColor(theme.modalBackground), theme.modalBackground]}
-                    />
-                )}
-                {hasRouteError && (
-                    <DotIndicatorMessage
-                        style={[styles.mh5, styles.mv3]}
-                        messages={ErrorUtils.getLatestErrorField(transaction, 'route')}
-                        type="error"
-                    />
-                )}
-            </View>
+    const updateWaypoints = useCallback(
+        ({data}) => {
+            if (_.isEqual(waypointsList, data)) {
+                return;
+            }
+
+            const newWaypoints = {};
+            _.each(data, (waypoint, index) => {
+                const newWaypoint = lodashGet(waypoints, waypoint, {});
+                newWaypoints[`waypoint${index}`] = lodashIsEmpty(newWaypoint) ? null : newWaypoint;
+            });
+
+            setOptimisticWaypoints(newWaypoints);
+            // eslint-disable-next-line rulesdir/no-thenable-actions-in-views
+            Transaction.updateWaypoints(transactionID, newWaypoints).then(() => {
+                setOptimisticWaypoints(null);
+            });
+        },
+        [transactionID, waypoints, waypointsList],
+    );
+
+    const renderItem = ({item, drag, getIndex, isActive}) => {
```

Comment:
can you extract this as a curry function and extract it outside the `DistanceRequest` component? I believe it would make it much easier to see which variables are coming form the local state of `DistanceRequest`.

Reply:
and use `onDragEnd={() => updateWaypoints({transactionID, waypoints, waypointList})}`?

Reply:
Can you convert this render function into a component and define that component outside `DistanceRequest` component?

In addition to `item, drag, getIndex, isActive`, I think we only have to take `navigateToWaypointEditPage` as a `onPress` prop and `isLoadingRoute` as a `disabled` prop.